syntax = "proto2";
package DB.Protos;

import "data_models.proto";
import "cnch_common.proto";

option cc_generic_services = true;

message ExecuteSimpleQueryReq
{
  required string query = 1;
  optional string query_id = 2;
  optional string settings = 3;
  optional ClientInfo client_info = 4;
}

message ExecuteSimpleQueryResp
{
  optional string exception = 1;
  required uint32 server_revision = 2;
  optional string block = 3;
}

message SubmitManipulationTaskReq
{
  required uint64 txn_id = 1;
  required uint64 timestamp = 2;

  /// ManipulationTaskParams
  required uint32 type = 3;
  required string task_id = 4;
  required uint32 rpc_port = 5;
  required uint64 columns_commit_time = 6;

  optional string create_table_query = 7;
  repeated DataModelPart source_parts = 8;

  /// Mutation task
  optional uint64 mutation_commit_time = 9;
  optional string mutate_commands = 10;
  /// used to mark task of bucket table
  optional bool is_bucket_table = 11;
}

message SubmitManipulationTaskResp
{
  optional string exception = 1;
}

message SubmitKafkaConsumeTaskReq
{
  required uint32 type = 1;
  required string task_id = 2;
  required uint32 rpc_port = 3;

  required string database = 4;
  required string table = 5;
  optional StorageID cnch_storage_id = 6;
  required uint64 assigned_consumer = 7;
  repeated string create_table_command = 8;
  repeated TopicPartitionModel tpl = 9;
}

message SubmitKafkaConsumeTaskResp
{
  optional string exception = 1;
}

message GetConsumerStatusReq
{
  required StorageID table = 1;
}

message GetConsumerStatusResp
{
  optional string exception = 1;
  optional string cluster = 2;
  repeated string topics = 3;
  repeated string assignments = 4;
  optional uint32 consumer_num = 5;
  optional string last_exception = 6;
}

message PreloadChecksumsAndPrimaryIndexReq
{
  required StorageID table = 1;
}

message PreloadChecksumsAndPrimaryIndexResp
{
  optional string exception = 1;
}

message ShutdownManipulationTasksReq
{
  required UUID table_uuid = 1;
  repeated string task_ids = 2;
}

message ShutdownManipulationTasksResp
{
  optional string exception = 1;
}

message TouchManipulationTasksReq
{
  required UUID table_uuid = 1;
  repeated string tasks_id = 2;
}

message TouchManipulationTasksResp
{
  optional string exception = 1;
  repeated string tasks_id = 2;
}

message ManipulationTask
{
  required string task_id = 1;
  optional uint32 type = 2;
  optional StorageID storage_id = 3;
  optional float elapsed = 4;
  optional int64 num_parts = 5;
  repeated string source_part_names = 6;
  repeated string result_part_names = 7;
  optional string partition_id = 8;
  optional int64 total_size_bytes_compressed = 9;
  optional int64 total_size_marks = 10;
  optional float progress = 11;
  optional int64 bytes_read_uncompressed = 12;
  optional int64 bytes_written_uncompressed = 13;
  optional int64 rows_read = 14;
  optional int64 rows_written = 15;
  optional int64 columns_written = 16;
  optional int64 memory_usage = 17;
  optional int64 thread_id = 18;
}

message GetManipulationTasksStatusReq
{
}

message GetManipulationTasksStatusResp
{
  optional string exception = 1;
  repeated ManipulationTask tasks = 2;
}

message GetPreallocatedStatusReq
{
  required UUID storage_id = 1;
}

message GetPreallocatedStatusResp
{
  optional string exception = 1;
  required bool preallocated = 2;
}

message SetQueryIntentReq
{
  required UUID storage_id = 1;
  required string query_id = 2;
  required uint64 start_time = 3;
  required uint64 worker_topology_hash = 4;
}

message SetQueryIntentResp
{
  optional string exception = 1;
  required bool preallocated = 2;
  required uint32 status = 3;
}

message PreloadDataPartsReq
{
  optional uint64 txn_id = 1;
  optional string create_table_query = 2;
  repeated DataModelPart parts = 3;
  optional bool sync = 4;
}

message PreloadDataPartsResp
{
  optional string exception = 1;
}

message SubmitSyncTaskReq
{
  required UUID storage_id = 1;
  required string database = 2;
  required string table = 3;
  required string query_id = 4;
  required uint32 trigger_type = 5;
  required bool full_sync = 6;
  repeated DataModelPart parts = 7;
  optional uint64 worker_topology_hash = 8;
  required uint32 rpc_port = 9;
  required string create_table_query = 10;
}

message SubmitSyncTaskResp
{
  optional string exception = 1;
}

message ResetQueryIntentReq
{
  required UUID storage_id = 1;
  required string query_id = 2;
}

message ResetQueryIntentResp
{
  optional string exception = 1;
}

message SubmitScaleTaskReq
{
  required UUID storage_id = 1;
  required string database = 2;
  required string table = 3;
  required uint64 worker_topology_hash = 4;
  required uint32 rpc_port = 5;
  required string create_table_query = 6;
}

message SubmitScaleTaskResp
{
  optional string exception = 1;
}

message GetCloudMergeTreeStatusReq
{
  required StorageID storage_id = 1;
}

message GetCloudMergeTreeStatusResp
{
  optional string exception = 1;
  optional bool table_exist = 2;
}

message ClearPreallocatedDataPartsReq
{
  required string database = 1;
  required string table = 2;
  required string create_table_query = 3;
  repeated string parts_to_remove = 4;
}

message ClearPreallocatedDataPartsResp
{
  optional string exception = 1;
}

message CreateDedupWorkerReq
{
  required StorageID table = 1;
  required string create_table_query = 2;
  required HostWithPorts host_ports = 3;
}

message CreateDedupWorkerResp
{
  optional string exception = 1;
}

message DropDedupWorkerReq
{
  required StorageID table = 1;
}

message DropDedupWorkerResp
{
  optional string exception = 1;
}

message GetDedupWorkerStatusReq
{
  required StorageID table = 1;
}

message GetDedupWorkerStatusResp
{
  optional string exception = 1;
  required bool is_active = 2;
  optional int64 create_time = 3;
  optional int64 total_schedule_cnt = 4;
  optional int64 total_dedup_cnt = 5;
  optional int64 last_schedule_wait_ms = 6;
  optional int64 last_task_total_cost_ms = 7;
  optional int64 last_task_dedup_cost_ms = 8;
  optional int64 last_task_publish_cost_ms = 9;
  optional int64 last_task_staged_part_cnt = 10;
  optional int64 last_task_visible_part_cnt = 11;
  optional int64 last_task_staged_part_total_rows = 12;
  optional int64 last_task_visible_part_total_rows = 13;
  optional string last_exception = 14;
  optional int64 last_exception_time = 15;
}

message SendCreateQueryReq
{
  required uint64 txn_id = 1;
  required uint64 primary_txn_id = 2;
  required uint64 timeout = 3;
  repeated string create_queries = 4;
}

message SendCreateQueryResp
{
  optional string exception = 1;
}

message UDFInfo
{
  required string function_name = 1;
  required int64 version = 2;
}

message SendDataPartsReq
{
  required uint64 txn_id = 1;
  required string database_name = 2;
  required string table_name = 3;
  repeated DataModelPart parts = 4;
  repeated DataModelDeleteBitmap bitmaps = 5;
  repeated int64 bucket_numbers = 6;
  repeated UDFInfo udf_infos = 7;
  optional string disk_cache_mode = 8;
}

message SendDataPartsResp
{
  optional string exception = 1;
}

message SendCnchHiveDataPartsReq
{
  required uint64 txn_id = 1;
  required string database_name = 2;
  required string table_name = 3;
  repeated CnchHivePartModel parts = 4;
}

message SendCnchHiveDataPartsResp
{
  optional string exception = 1;
}

message CheckDataPartsReq
{
  required uint64 txn_id = 1;
  required string database_name = 2;
  required string table_name = 3;
  required string create_query = 4;
  repeated DataModelPart parts = 5;
}

message CheckDataPartsResp
{
  optional string exception = 1;
  repeated string part_path = 2;
  repeated bool is_passed = 3;
  repeated string message = 4;
}

message OffloadingTableResource
{
  required UUID table_id = 1;
  required string worker_table_name = 2;
  repeated HostWithPorts buffer_workers = 3;
}

message SendOffloadingReq
{
  required uint64 txn_id = 1;
  repeated HostWithPorts read_workers = 2;
  repeated OffloadingTableResource table_resource = 3;
}

message SendOffloadingResp
{
  optional string exception = 1;
}

message TableDataParts
{
  required string database = 1;
  required string table = 2;
  repeated DataModelPart server_parts = 3;
  repeated DataModelDeleteBitmap server_part_bitmaps = 4;
  repeated CnchHivePartModel hive_parts = 5;
  repeated int64 bucket_numbers = 6;
}

message SendResourcesReq
{
  required uint64 txn_id = 1;
  required uint64 primary_txn_id = 2;
  required uint64 timeout = 3;
  /// create queries
  repeated string create_queries = 4;
  /// data parts
  repeated TableDataParts data_parts = 5;
  optional string disk_cache_mode = 6;
}

message SendResourcesResp
{
  optional string exception = 1;
}

message RemoveWorkerResourceReq
{
  required uint64 txn_id = 1;
}

message RemoveWorkerResourceResp
{
  optional string exception = 1;
}

service CnchWorkerService
{
  rpc executeSimpleQuery(ExecuteSimpleQueryReq) returns (ExecuteSimpleQueryResp);

  // ManipulationTasks
  rpc submitManipulationTask(SubmitManipulationTaskReq) returns (SubmitManipulationTaskResp);
  rpc touchManipulationTasks(TouchManipulationTasksReq) returns (TouchManipulationTasksResp);
  rpc shutdownManipulationTasks(ShutdownManipulationTasksReq) returns (ShutdownManipulationTasksResp);
  rpc getManipulationTasksStatus(GetManipulationTasksStatusReq) returns (GetManipulationTasksStatusResp);

  rpc submitKafkaConsumeTask(SubmitKafkaConsumeTaskReq) returns (SubmitKafkaConsumeTaskResp);
  rpc getConsumerStatus(GetConsumerStatusReq) returns (GetConsumerStatusResp);

  rpc preloadChecksumsAndPrimaryIndex(PreloadChecksumsAndPrimaryIndexReq) returns (PreloadChecksumsAndPrimaryIndexResp);

  rpc GetPreallocatedStatus(GetPreallocatedStatusReq) returns (GetPreallocatedStatusResp);
  rpc SetQueryIntent(SetQueryIntentReq) returns (SetQueryIntentResp);
  rpc SubmitSyncTask(SubmitSyncTaskReq) returns (SubmitSyncTaskResp);
  rpc ResetQueryIntent(ResetQueryIntentReq) returns (ResetQueryIntentResp);
  rpc SubmitScaleTask(SubmitScaleTaskReq) returns (SubmitScaleTaskResp);

  rpc getCloudMergeTreeStatus(GetCloudMergeTreeStatusReq) returns (GetCloudMergeTreeStatusResp);

  rpc ClearPreallocatedDataParts(ClearPreallocatedDataPartsReq) returns (ClearPreallocatedDataPartsResp);

  rpc createDedupWorker(CreateDedupWorkerReq) returns (CreateDedupWorkerResp);
  rpc dropDedupWorker(DropDedupWorkerReq) returns (DropDedupWorkerResp);
  rpc getDedupWorkerStatus(GetDedupWorkerStatusReq) returns (GetDedupWorkerStatusResp);

  rpc sendCreateQuery(SendCreateQueryReq) returns (SendCreateQueryResp);
  rpc sendQueryDataParts(SendDataPartsReq) returns (SendDataPartsResp);
  rpc sendCnchHiveDataParts(SendCnchHiveDataPartsReq) returns (SendCnchHiveDataPartsResp);
  rpc checkDataParts(CheckDataPartsReq) returns (CheckDataPartsResp);
  rpc sendOffloading(SendOffloadingReq) returns (SendOffloadingResp);
  rpc sendResources(SendResourcesReq) returns (SendResourcesResp);
  rpc removeWorkerResource(RemoveWorkerResourceReq) returns (RemoveWorkerResourceResp);
  rpc preloadDataParts(PreloadDataPartsReq) returns (PreloadDataPartsResp);
}
